sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -406,34 +464,47 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
      */
     public DatastoreV1.Read withNumQuerySplits(int numQuerySplits) {
       return toBuilder()
-          .setNumQuerySplits(Math.min(Math.max(numQuerySplits, 0), NUM_QUERY_SPLITS_MAX))
+          .setNumQuerySplits(StaticValueProvider.of(Math.min(Math.max(numQuerySplits, 0),
+              NUM_QUERY_SPLITS_MAX)))
           .build();
     }
 
+    public DatastoreV1.Read withNumQuerySplits(ValueProvider<Integer> numQuerySplits) {
+      return toBuilder().setNumQuerySplits(numQuerySplits).build();
+    }
+
     @Override
     public PCollection<Entity> expand(PBegin input) {
-      V1Options v1Options = V1Options.from(getProjectId(), getQuery(),
-          getNamespace());
+      V1Options v1Options = V1Options.from(getProjectId(), getNamespace());
 
       /*
        * This composite transform involves the following steps:
-       *   1. Create a singleton of the user provided {@code query} and apply a {@link ParDo} that
-       *   splits the query into {@code numQuerySplits} and assign each split query a unique
-       *   {@code Integer} as the key. The resulting output is of the type
-       *   {@code PCollection<KV<Integer, Query>>}.
+       *   1. Create a singleton of the user provided {@code query} or {@code gqlQuery} and
+       *   apply a {@link ParDo} that translates the {@code gqlQuery} into a {@query}, if a
+       *   gqlQuery is provided, else the query is bypassed.
+       *
+       *   2. A {@link ParDo} splits the resulting query into {@code numQuerySplits} and
+       *   assign each split query a unique {@code Integer} as the key. The resulting output is
+       *   of the type {@code PCollection<KV<Integer, Query>>}.
        *
        *   If the value of {@code numQuerySplits} is less than or equal to 0, then the number of
        *   splits will be computed dynamically based on the size of the data for the {@code query}.
        *
-       *   2. The resulting {@code PCollection} is sharded using a {@link GroupByKey} operation. The
+       *   3. The resulting {@code PCollection} is sharded using a {@link GroupByKey} operation. The
        *   queries are extracted from they {@code KV<Integer, Iterable<Query>>} and flattened to
        *   output a {@code PCollection<Query>}.
        *
-       *   3. In the third step, a {@code ParDo} reads entities for each query and outputs
+       *   4. In the third step, a {@code ParDo} reads entities for each query and outputs
        *   a {@code PCollection<Entity>}.
        */
+
+      Coder<KV<ValueProvider<String>, ValueProvider<Query>>> valueProviderCoder =
+          SerializableCoder.of(
+              new TypeDescriptor<KV<ValueProvider<String>, ValueProvider<Query>>>() {});
+
       PCollection<KV<Integer, Query>> queries = input
-          .apply(Create.of(getQuery()))
+          .apply(Create.of(KV.of(getGqlQuery(), getQuery())).withCoder(valueProviderCoder))
```

Comment:
Using KV here seems wrong -- it's not actually a key, right? Could we introduce some other type, possibly AutoValue and use a reasonable coder instead?